x-pack/auditbeat/module/system/login/login.go

```diff
@@ -0,0 +1,193 @@
+	IP        *net.IP
+	Timestamp time.Time
+	Origin    string
+}
+
+func (login LoginRecord) toMapStr() common.MapStr {
+	mapstr := common.MapStr{}
+
+	// Very useful for development
+	//mapstr.Put("utmp", fmt.Sprintf("%v", login.Utmp))
+
+	if login.TTY != "" {
+		mapstr.Put("tty", login.TTY)
+	}
+
+	return mapstr
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet collects login records from /var/log/wtmp.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config     Config
+	log        *logp.Logger
+	utmpReader *UtmpFileReader
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := defaultConfig
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
+	}
+
+	ms.utmpReader, err = NewUtmpFileReader(ms.log, bucket, config.UtmpFilePattern)
+	if err != nil {
+		return nil, err
+	}
+
+	return ms, nil
+}
+
+// Close cleans up the MetricSet when it finishes.
+func (ms *MetricSet) Close() error {
+	return ms.utmpReader.Close()
+}
+
+// Fetch collects any new login records from /var/log/wtmp. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	loginRecords, err := ms.utmpReader.ReadNew()
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+		return
+	}
+	ms.log.Debugf("%d new login records.", len(loginRecords))
+
+	for _, loginRecord := range loginRecords {
+		report.Event(ms.loginEvent(loginRecord))
+	}
+
+	// Save new state to disk
+	if len(loginRecords) > 0 {
+		err := ms.utmpReader.saveStateToDisk()
+		if err != nil {
+			ms.log.Error(err)
+			report.Error(err)
+			return
+		}
+	}
+}
+
+func (ms *MetricSet) loginEvent(loginRecord LoginRecord) mb.Event {
+	event := mb.Event{
+		Timestamp: loginRecord.Timestamp,
+		RootFields: common.MapStr{
+			"event": common.MapStr{
+				"type":   eventTypeEvent,
+				"action": loginRecord.Type,
+				"origin": loginRecord.Origin,
+			},
+		},
+		MetricSetFields: loginRecord.toMapStr(),
+	}
+
+	if loginRecord.Username != "" {
+		event.RootFields.Put("user.name", loginRecord.Username)
+
+		if loginRecord.UID != -1 {
+			event.RootFields.Put("user.id", loginRecord.UID)
+		}
+	}
+
+	if loginRecord.PID != -1 {
+		event.RootFields.Put("process.pid", loginRecord.PID)
+	}
+
+	if loginRecord.IP != nil {
+		event.RootFields.Put("source.ip", loginRecord.IP)
+	}
+
+	if loginRecord.Hostname != "" && loginRecord.Hostname != loginRecord.IP.String() {
+		event.RootFields.Put("source.hostname", loginRecord.Hostname)
```

Comment:
To be sure, this is the hostname from where the user logs in remotely, right? It's not the host on which the user logs in.

I wonder if `source.hostname` is the correct ECS field for this. It doesn't show up explicitly in ECS (only `source.ip`), but it does seem to make sense to me.

Reply:
Correct, it's the remote hostname. Since we already have `source.ip` I thought it makes sense. There's even `[source|destination].domain` in ECS already. 

I think other data sources could use this as well, e.g. flows where there might be a `source.hostname` and a `destination.hostname`.

Reply:
@andrewkroh Because the actual hostname might not be what another host sees? E.g. `webserver001` might be the hostname, but to the outside world it appears as `web.elastic.co`?